### ISSUE TITLE
Assembly descriptor removed but still in documentation

### ISSUE THREAD

**Issue by msillence** [eclipse/jkube#497]
### Description

descriptor support has been removed from the plugin but is still documented at the end of section 5.2.5 https://www.eclipse.org/jkube/docs/kubernetes-maven-plugin

This commit removed support:
https://github.com/eclipse/jkube/commit/cd475427f59c7df731e57b0516d4e8df0a0e9c3a#diff-7fe446a69cdae536171e449f96f5705c613db53de41eb9cddf032229e6e05893

### Info

* Eclipse JKube version : 1.0.2
* Maven version (`mvn -v`) : 3.0.6
```

```

* If it's a bug, how to reproduce : 

try and build a docker image using example documentation and you get the error:

Unable to parse configuration of mojo org.eclipse.jkube:kubernetes-maven-plugin:1.0.2:build for parameter descriptors: Cannot find 'descriptors' in class org.eclipse.jkube.kit.common.AssemblyConfiguration

**Comment by rohanKanojia** [comment]
Which descriptors element are you referring to? I'm not able to find it in assembly configuration docs.

**Comment by msillence** [comment]
`  <assembly>
    <mode>dir</mode>
    <targetDir>/opt/demo</targetDir>
    <descriptor>assembly.xml</descriptor>
  </assembly>
</build>`

just above 5.2.6. Assembly

I was hoping for a drop in replacement but I was using this feature in fabirc8

**Comment by rohanKanojia** [comment]
ah, I'm sorry. Looks like we missed it. This is not supported anymore since we refactored jkube-kit out of maven. Could you please elaborate what you're trying to package into your Docker image using this `assembly.xml`? Maybe we can provide you an alternative..

**Comment by manusa** [comment]
The only option right now is for you to inline the descriptor in the configuration

**Comment by msillence** [comment]
My assembly is:
```
<inline>

	<fileSets>
		<fileSet>
			<outputDirectory>/</outputDirectory>
			<directory>src/main/resources/docker/</directory>
			<lineEnding>unix</lineEnding>
			<fileMode>0755</fileMode>
		</fileSet>
	</fileSets>

	<dependencySets>
		<dependencySet>
			<scope>provided</scope>
			<outputDirectory>.</outputDirectory>
		</dependencySet>

		<dependencySet>
			<scope>compile</scope>
			<outputDirectory>.</outputDirectory>
		</dependencySet>
	</dependencySets>
</inline>
```
it says lineEnding and dependencySets are not supported

**Comment by manusa** [comment]
For this more complex scenarios, I would recommend using the maven-resources-plugin or maven-assembly-plugin first to preprocess and then use the target directory as the fileSet source for JKube.

Check out an example here: http://blog.marcnuri.com/quarkus-jkube-qute-markdown-different-sources (The `Build: Maven Assembly Plugin` describes a setup similar to what you need)

So. you can keep your assembly.xml descriptor, but you'll need to configure it the maven-assembly-plugin and run the assembly prior to kubernetes-maven-plugin build task.

**Comment by msillence** [comment]
thanks, that worked, though I couldnt get it working with a Dockerfile, the docker file is so minimal I've converted it to xml and embedded it in the pom.

I'm not sure if I'm allowed to mix a dockerfile and an assembly/inline - it's probably my lack of understanding about the copy path and using the directory assembly

**Comment by manusa** [comment]
> I'm not sure if I'm allowed to mix a dockerfile and an assembly/inline - it's probably my lack of understanding about the copy path and using the directory assembly

Yes you should be able

Take a look at this quickstart https://github.com/eclipse/jkube/tree/master/quickstarts/maven/docker-file-provided which showcases different possibilities to work with external Docker files.

I think you're probably interested in this one:
https://github.com/eclipse/jkube/blob/db9c522bce061f4e327c4e9f06930b5ac29b70b2/quickstarts/maven/docker-file-provided/pom.xml#L149

**Comment by msillence** [comment]
OK this works for me:
			<plugin>
				<artifactId>maven-assembly-plugin</artifactId>
				<version>3.3.0</version>
				<configuration>
					<appendAssemblyId>false</appendAssemblyId>
					<finalName>preassembled</finalName>
					<outputDirectory>target/assembly</outputDirectory>
					<descriptors>
						<descriptor>src/main/assembly/assembly.xml</descriptor>
					</descriptors>
				</configuration>
...
			</plugin>

			<plugin>
				<groupId>org.eclipse.jkube</groupId>
				<artifactId>kubernetes-maven-plugin</artifactId>
				<version>1.0.2</version>
				<configuration>
					<images>
						<image>
							<name>artifactname:${project.version}</name>

							<build>
								<contextDir>${project.basedir}/</contextDir>
								<dockerFile>src/main/docker/Dockerfile</dockerFile>
								<assembly>
									<name>target/assembly/preassembled</name>
								</assembly>
							</build>
						</image>
					</images>
				</configuration>

Note that the contextDir is common to both the path to the dockerfile and the assembly directory
The Dockerfile copy must then match the images/image/build/assembly/name

`COPY --chown=app:app target/assembly/preassembled/ /app/`

If you don't have a name configured in the assembly then you get a null pointer in the plugin

I feel this is now really off topic to the orignal request

**Comment by manusa** [comment]
> I feel this is now really off topic to the original request

Yes, we still need to fix the documentation or point to a workaround.

**Comment by rohanKanojia** [comment]
@msillence : Hi, Would it be possible for you to contribute a PR to fix documentation and provide workaround?

**Comment by msillence** [comment]
When I moved this config to my real project it failed with an OOM - it was recursively adding the entire target folder to the docker image build in the target folder - until it failed

This is my current config that doesn't suffer from that problem - not sure why it was working on the smaller test project: 

```
							<build>
								<contextDir>${project.basedir}/target/assembly/preassembled</contextDir>
								<dockerFile>${project.basedir}/src/main/docker/Dockerfile</dockerFile>
								<assembly>
									<name>./</name>
								</assembly>
							</build>
```

and docker file now reads:

`COPY ./ /app/`

I'll have a think about the PR assuming that my workaround is the right approach

**Comment by rohanKanojia** [comment]
hmmm, I'm afraid if this could be a bug in our plugin itself. Would it be possible for you to share a demo project so that we can reproduce this OOM error?

**Comment by rohanKanojia** [comment]
I think OOM error might be related to https://github.com/eclipse/jkube/issues/561 https://github.com/eclipse/jkube/issues/544  https://github.com/eclipse/jkube/issues/529 (I have a PR #567 that should fix #529)

I'm removing `<descriptor>` reference from the documentation
